File: app/src/main/java/org/stepic/droid/services/DeleteService.java
package org.stepic.droid.services;

import android.app.DownloadManager;
import android.app.IntentService;
import android.content.Intent;
import android.util.Log;

import com.squareup.otto.Bus;
import com.yandex.metrica.YandexMetrica;

import org.stepic.droid.base.MainApplication;
import org.stepic.droid.model.Course;
import org.stepic.droid.model.Lesson;
import org.stepic.droid.model.Section;
import org.stepic.droid.model.Step;
import org.stepic.droid.model.Unit;
import org.stepic.droid.preferences.UserPreferences;
import org.stepic.droid.store.IStoreStateManager;
import org.stepic.droid.store.operations.DatabaseManager;
import org.stepic.droid.util.AppConstants;
import org.stepic.droid.util.resolvers.IVideoResolver;
import org.stepic.droid.web.IApi;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

public class DeleteService extends IntentService {
    @Inject
    DownloadManager mSystemDownloadManager;
    @Inject
    UserPreferences mUserPrefs;
    @Inject
    Bus mBus;
    @Inject
    IVideoResolver mResolver;
    @Inject
    IApi mApi;
    @Inject
    DatabaseManager mDb;
    @Inject
    IStoreStateManager mStoreStateManager;

    public DeleteService() {
        super("delete_service");

    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        MainApplication.component().inject(this);
        super.onStartCommand(intent, flags, startId);
        return START_REDELIVER_INTENT;
    }


    @Override
    protected void onHandleIntent(Intent intent) {
        LoadService.LoadTypeKey type = (LoadService.LoadTypeKey) intent.getSerializableExtra(AppConstants.KEY_LOAD_TYPE);
        Log.i("downloading", "Service: " + Thread.currentThread().getName());
        try {
            switch (type) {
                case Course:
                    Course course = (Course) intent.getSerializableExtra(AppConstants.KEY_COURSE_BUNDLE);
                    DatabaseManager.Table tableType = (DatabaseManager.Table) intent.getSerializableExtra(AppConstants.KEY_TABLE_TYPE);
                    removeFromDisk(course, tableType);
                    break;
                case Section:
                    Section section = (Section) intent.getSerializableExtra(AppConstants.KEY_SECTION_BUNDLE);
                    removeFromDisk(section);
                    break;
                case UnitLesson:
                    Unit unit = (Unit) intent.getSerializableExtra(AppConstants.KEY_UNIT_BUNDLE);
                    Lesson lesson = (Lesson) intent.getSerializableExtra(AppConstants.KEY_LESSON_BUNDLE);
                    removeFromDisk(unit, lesson);
                    break;
                case Step:
                    Step step = (Step) intent.getSerializableExtra(AppConstants.KEY_STEP_BUNDLE);
                    removeFromDisk(step);
                    break;
            }
        } catch (NullPointerException ex) {
            //possibly user click clear cache;
//            throw ex;
            YandexMetrica.reportError("DeleteService nullptr", ex);

            mDb.dropDatabase();
        }
    }

    private void removeFromDisk(Step step) {
        if (step.getBlock().getVideo() != null) {
            String path = mDb.getPathToVideoIfExist(step.getBlock().getVideo());
            File file = new File(path);
            if (file.exists()) {
                file.delete();
            }

            mDb.deleteVideo(step.getBlock().getVideo());
        }

        mDb.deleteStep(step); // remove steps
        mStoreStateManager.updateStepAfterDeleting(step);
    }

    private void removeFromDisk(Unit unit, Lesson lesson) {
        List<Step> steps = mDb.getStepsOfLesson(lesson.getId());
        for (Step step : steps) {
            removeFromDisk(step);
        }

//        unit.setIs_cached(false);
//        unit.setIs_loading(false);
//        lesson.setIs_cached(false);
//        lesson.setIs_loading(false);
//        mDb.updateOnlyCachedLoadingLesson(lesson);
//        mDb.updateOnlyCachedLoadingUnit(unit);
//        mStoreStateManager.updateUnitLessonAfterDeleting(unit);

    }

    private void removeFromDisk(Section section) {
        List<Unit> units = mDb.getAllUnitsOfSection(section.getId());
        List<Lesson> lessons = new ArrayList<>();
        for (Unit unit : units) {
            lessons.add(mDb.getLessonOfUnit(unit));
        }

        for (int i = 0; i < units.size(); i++) {
            removeFromDisk(units.get(i), lessons.get(i));
        }


//        mDb.deleteSection(section);
    }

    private void removeFromDisk(Course course, DatabaseManager.Table tableType) {
        List<Section> sections = mDb.getAllSectionsOfCourse(course);
        for (Section section : sections) {
            removeFromDisk(section);
        }
//        mDb.deleteCourse(course, tableType);
    }
}

File: app/src/main/java/org/stepic/droid/services/DeleteService.kt
package org.stepic.droid.services

import android.app.DownloadManager
import android.app.IntentService
import android.app.Service
import android.content.Intent
import android.os.Handler
import org.stepic.droid.analytic.Analytic
import org.stepic.droid.base.App
import org.stepic.droid.core.downloads.contract.DownloadsPoster
import org.stepik.android.model.structure.Lesson
import org.stepik.android.model.structure.Step
import org.stepic.droid.storage.StoreStateManager
import org.stepic.droid.storage.operations.DatabaseFacade
import org.stepic.droid.util.AppConstants
import org.stepic.droid.util.SuppressFBWarnings
import java.io.File
import javax.inject.Inject

class DeleteService : IntentService("delete_service") {
    @Inject
    lateinit var systemDownloadManager: DownloadManager
    @Inject
    lateinit var databaseFacade: DatabaseFacade
    @Inject
    lateinit var storeStateManager: StoreStateManager
    @Inject
    lateinit var analytic: Analytic

    @Inject
    lateinit var downloadsPoster: DownloadsPoster

    override fun onCreate() {
        super.onCreate()
        App
                .componentManager()
                .downloadsComponent()
                .inject(this)
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        super.onStartCommand(intent, flags, startId)
        return Service.START_REDELIVER_INTENT
    }


    override fun onHandleIntent(intent: Intent) {
        val type = intent.getSerializableExtra(AppConstants.KEY_LOAD_TYPE) as? LoadService.LoadTypeKey
        try {
            when (type) {
                LoadService.LoadTypeKey.Section -> {
                    val sectionId = intent.getLongExtra(AppConstants.KEY_SECTION_BUNDLE, -1).takeIf { it >= 0 } ?: return
                    removeFromDisk(sectionId)
                }
                LoadService.LoadTypeKey.Lesson -> {
                    val lesson = intent.getParcelableExtra<Lesson>(AppConstants.KEY_LESSON_BUNDLE)
                    removeFromDisk(lesson)
                }
                LoadService.LoadTypeKey.Step -> {
                    val step = intent.getSerializableExtra(AppConstants.KEY_STEP_BUNDLE) as? Step
                    removeFromDisk(step)
                }
            }
        } catch (ex: NullPointerException) {
            //possibly user click clear cache;
            //            throw ex;
            analytic.reportError(Analytic.Error.DELETE_SERVICE_ERROR, ex)
            databaseFacade.dropDatabase()
        }

    }

    @SuppressFBWarnings(value = "RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE")
    private fun removeFromDisk(step: Step?) {
        step?.block?.video?.let {
            val cachedVideo = databaseFacade.getCachedVideoIfExist(it)
            var file = File(cachedVideo?.url)
            if (cachedVideo != null && file.exists()) {
                file.delete()
            }

            //delete png thumbnail
            file = File(cachedVideo?.thumbnail)
            if (cachedVideo != null && file.exists()) {
                file.delete()
            }

            databaseFacade.deleteVideo(it)
        }
        step?.let {
            step.isCached = false
            step.isLoading = false
            databaseFacade.updateOnlyCachedLoadingStep(step)
//            database.deleteStep(step) // remove steps FIXME: MAYBE NOT DELETE STEP?
            storeStateManager.updateStepAfterDeleting(step)
            val mainHandler = Handler(App.getAppContext().mainLooper)
            mainHandler.post {
                downloadsPoster.stepRemoved(step.id)
            }
        }
    }

    @SuppressFBWarnings(value = "RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE")
    private fun removeFromDisk(lesson: Lesson?) {
        if (lesson == null) {
            return
        }
        val steps = databaseFacade.getStepsOfLesson(lesson.id)
        for (step in steps) {
            removeFromDisk(step)
        }
    }

    @SuppressFBWarnings(value = "RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE")
    private fun removeFromDisk(sectionId: Long) {
        val units = databaseFacade.getAllUnitsOfSection(sectionId)
        val lessons = units
                .mapNotNull { databaseFacade.getLessonOfUnit(it) }

        for (lesson in lessons) {
            removeFromDisk(lesson)
        }
    }
}
